### ISSUE TITLE
WasmImportLinkage is not propagated in LibraryImport

### ISSUE THREAD

**Issue by RReverser** [dotnet/runtime#109181]
### Description

As per https://github.com/dotnet/runtime/issues/93824#issuecomment-1774067140:

> > Not really API Review related, but would we want to have LibraryImport automatically propagate this attribute to the inner DllImport? My initial thought is yes, but just wanted to cover my bases.
> 
> Correct, this would need to be propagated.

This works as expected for simple functions that don't require any transform by the `LibraryImport`, but seems to be broken when using any marshalled types.

### Reproduction Steps

Example:

```cs
[LibraryImport("foo")]
[WasmImportLinkage]
public static partial void test1();

[LibraryImport("foo")]
[WasmImportLinkage]
public static partial void test2(byte[] marshalMe);
```

### Expected behavior

Both functions become Wasm imports.

### Actual behavior

Only simple `test1` function becomes a Wasm import. The generated `pinvoke-table.h` looks like this:

```c
__attribute__((import_module("foo"),import_name("test1")))
extern void SpacetimeDB_Internal_23_foo_23_test1 ();

void test2 (void *);
```

This, in turn, results in a linker error:

```
    wasm-ld : error : D:/.../obj/Release/net9.0/wasi-wasm/wasm/for-publish/pinvoke.o: undefined symbol: test2
```

### Known Workarounds

1. Use `DllImport` with manual marshalling, which can get pretty cumbersome when using a lot of functions.
2. Manually declare all functions in an extra C file with `__attribute__((import_module("foo"))` applied to each of them, and add it to MSBuild linker args. This is what we did before .NET 9 as `WasmImportLinkage` didn't exist yet, but it is also pretty cumbersome to maintain such a list manually.

**Comment by RReverser** [comment]
cc @lewing as you worked on the original attribute

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky Are we not propogating `WasmImportLinkage`?

**Comment by jkoritzinsky** [comment]
I don't think we ever added support for that. We definitely should.

**Comment by RReverser** [comment]
Other than a guess mentioned in the issue description, I don't really understand why `test1` does work as expected.

It seems the PInvokeCollector only considers functions marked with DllImport, so if propagation isn't implemented... how come those functions work?

Do they get simply copied with all the original attributes if they don't use marshaling?

**Comment by jkoritzinsky** [comment]
For functions with no marshalling, the `LibraryImport` generator just puts a corresponding `DllImport` attribute on the same method (to avoid emitting extra code that's just forwarding to an inner function call)

**Comment by RReverser** [comment]
I was going to say, but how does that work if source generators can only add new items, not modify in-place, but I simply forgot that you can, in fact, add more attributes on partial method declarations of the same method.

That makes sense, but makes the behaviour pretty confusing to an unsuspecting user. A single added parameter suddenly breaking a function that was imported correctly before can be non-trivial to debug.

Given the propagation is not something that's going to be fixed for 9.0, I wonder if it would be possible to at least consistently forbid LibraryImport+UnmanagedCallersOnly for now, with diagnostic clearly saying it's unsupported? I think that would avoid quite a lot of confusion and bug reports in the year to come between .NET 9 and .NET 10.

**Comment by SingleAccretion** [comment]
> I wonder if it would be possible to at least consistently forbid LibraryImport+UnmanagedCallersOnly for now, with diagnostic clearly saying it's unsupported?

I would wager a diagnostic like this would be higher in risk than backporting the change proper.

**Comment by RReverser** [comment]
It sounds like the change is not implemented yet, so it's not simply backporting?

And since a fix for said diagnostic is just to replace [LibraryImport] with [DllImport] - as it would only show up on functions where it's safe to do so as no marshaling is required - it still seems like a much better / less confusing state than the current one where propagation works seemingly randomly from developer's PoV.

**Comment by SingleAccretion** [comment]
> It sounds like the change is not implemented yet, so it's not simply backporting?

Yes, I meant once it is implemented.

A diagnostic like this would require backporting a _separate_ change, which would then need to be reverted in main. I simply think it is better to service the proper change (and I would also be strongly favor of doing that, since this is effectively broken new .NET 9 functionality).

**Comment by RReverser** [comment]
> I simply think it is better to service the proper change (and I would also be strongly favor of doing that, since this is effectively broken new .NET 9 functionality).

Are you saying you think it would be possible to backport it without waiting for .NET 10? If so, I agree, that would be the best change.

I'm just somewhat sceptical because the original attribute landed very shortly after .NET 8 has been released, and there was a strong pushback to backporting it despite the PR author being in favour (see https://github.com/dotnet/runtime/pull/94615#issuecomment-1823892554). So, instead, we had to wait a year to get this attribute... and now it seems like we'd have to wait another year for it to be fixed.

But, rereading that discussion, I can see how this one could be treated as a bugfix and hopefully eligible for backporting. If so, looking forward to it 🤞 These attributes working in conjunction would remove a lot of custom forwarding C code for us.
